### PR TITLE
URA-872 - improve benchmarking

### PR DESCRIPTION
**Pull request by jethror1**
- improvements to benchmarking
  - switch to using [memory-profiler](https://pypi.org/project/memory-profiler/) for measuring memory usage, this more accurately measures total memory usage of all child processes
  - add `--repeats` to allow for running same benchmark multiple times and calculating mean
  - add more details to readme
- add fix to allow authentication using AWS SSO
  - after authenticating, if the environment variable `AWS_DEFAULT_PROFILE` is set to the login profile this is used for authentication and requests
  - TODO:
    - need to also handle taking access keys directly to be able to provide long term limited scope tokens
    - add details to the readme on authentication
-  fix some broken tests

- - -
This change is [Reviewable](https://reviewable.io/reviews/eastgenomics/s3_upload/36)

### CODE REVIEW

File: s3_upload/utils/upload.py [hunk 1]
```diff
@@ -213,12 +222,14 @@ def multi_thread_upload(
...
 
-    session = boto3.session.Session()
+    session = boto3.session.Session(profile_name=AWS_DEFAULT_PROFILE)
     s3_client = session.client(
         "s3",
         config=Config(
             retries={"total_max_attempts": 10, "mode": "standard"},
             disable_request_compression=True,
+            tcp_keepalive=True,
+            max_pool_connections=100,
```
Comment (fletchercr) [inline s3_upload/utils/upload.py:232]: Just curious, why is it set to 100? What does this allow?

**Review by jethror1 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/eastgenomics/s3_upload/36)* status: 0 of 8 files reviewed, 1 unresolved discussion (waiting on @fletchercr)

___
*[`s3_upload/utils/upload.py` line 232 at r1](https://reviewable.io/reviews/eastgenomics/s3_upload/36#gh-1832847699:-OB6FPiC-on-z_7K7h7F:b-uxc71) ([raw file](https://github.com/eastgenomics/s3_upload/blob/6a7d27b2b316a9787edca80596d2a9cbde6ea92d/s3_upload/utils/upload.py#L232)):*
<details><summary><i>Previously, fletchercr wrote…</i></summary><blockquote>

Just curious, why is it set to 100? What does this allow?
</blockquote></details>

sorry should have added detail on this, this is so that in the session that is split across threads it can handle more than the default of 10 connections at once, this is from the jobs:
```
max_pool_connections (int) – The maximum number of connections to keep in a connection pool. If this value is not set, the default value of 10 is used.
```

fairly sure if we were to use > 10 threads on a core, this not being set would limit it

File: README.md [hunk 2]
```diff
@@ -105,13 +105,18 @@ The expected fields in this log file are:
...
+A small [benchmarking script](https://github.com/eastgenomics/s3_upload/blob/main/scripts/benchmark.py) has been written to be able to repeatedly call the uploader with a set number of cores and threads at once to determine the optimal setting for upload time and available compute. It will iterate through combinations of the provided cores and threads, uploading a given run directory and automatically deleting the uploaded files on completion. Results are then written to a file `s3_upload_benchmark_{datetime}.tsv` in the current directory. This allows for measuring the total upload time and maximum resident set size (i.e. peak memory usage). This is using the [memory-profiler](https://pypi.org/project/memory-profiler/) package to measure combined memory usage of all spawned child processes to run the upload.
```
Comment (fletchercr) [inline README.md:108]: Also, just curious, why use memory-profiler instead of the previous method (stderr output of /usr/bin/time)?

**Review by jethror1 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/eastgenomics/s3_upload/36)* status: 0 of 8 files reviewed, 2 unresolved discussions (waiting on @fletchercr)

___
*[`README.md` line 108 at r1](https://reviewable.io/reviews/eastgenomics/s3_upload/36#gh-1832892749:-OB6Hj4NB2GXZdiiiF0w:b-d1av4n) ([raw file](https://github.com/eastgenomics/s3_upload/blob/6a7d27b2b316a9787edca80596d2a9cbde6ea92d/README.md#L108)):*
I realised that `usr/bin/time -v` does not handle measuring memory across child processes (i.e. when we use ProcessPoolExecutor and ThreadPoolExecutor it spawns a lot of additional child processes which aren't captured). We can easily show this by doing the following:

Define simple function to generate a 4 string variables of 1 million characters each across 4 separate processes:
```
from concurrent.futures import (
    ProcessPoolExecutor,
    as_completed,
)
from time import sleep


def test(x):
    x = x * 10000000
    sleep(2)


l = ["a", "b", "c", "d"]

with ProcessPoolExecutor(max_workers=8) as executor:

    concurrent_jobs = {executor.submit(test, x) for x in l}

    for future in as_completed(concurrent_jobs):
        # access returned output as each is returned in any order
        future.result()
```

This will allocate at least 9.5mb of memory each which we can check with the following:
```
>>> sys.getsizeof('a' * 10000000) / 1024 / 1024
9.536789894104004
```

If we save this to a `test.py` file and call it with both `/usr/bin/time -v` and `mprof` we can compare values.

With `/usr/bin/time -v` we get:
```
$ /usr/bin/time -v python3 test.py 2>&1 | grep Maximum
        Maximum resident set size (kbytes): 20152
```
This gives 20,152kb, which is only ~20Mb and therefore is definitely not enough for the overhead of Python 4 times and the 4 x ~9.5mb variables in memory.

Testing this with `mprof`:
```
jethro@jethro-T490:~/Projects/s3_upload$ mprof run --include-children -o mprof.out python3 test.py
mprof: Sampling memory every 0.1s
running new process
jethro@jethro-T490:~/Projects/s3_upload$ mprof peak mprof.out 
mprof.out       131.938 MiB
```

This amount of memory being used I can more believe, since it's 4 separate Python processes being run with that ~40mb of memory allocation for the variables, plus the over head of opening up a ProcessPoolExecutor.

Hopefully all that makes sense, it's not super important to be exactly correct since it's just for benchmarking, but I'd like it to be vaguely close.

**Review by fletchercr — APPROVED** [review]
> Reviewed 8 of 8 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/eastgenomics/s3_upload/36)* status: <span>:shipit:</span> complete! all files reviewed, all discussions resolved (waiting on @jethror1)

**Pull request merged**